main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+      case (Term.Cst(_), Term.True) => throw ConflictException(t1, t2, loc)
+      case (Term.Cst(_), Term.False) => throw ConflictException(t1, t2, loc)
+      case (Term.True, Term.Cst(_)) => throw ConflictException(t1, t2, loc)
+      case (Term.False, Term.Cst(_)) => throw ConflictException(t1, t2, loc)
+
+      // Non-trivial and non-conflicted equation: keep it.
+      case _ => Equation(t1, t2, loc) :: simplify(es)
+    }
+  }
+
+  /**
+    * Propagates constants and truth values through the equation system.
+    *
+    * The implementation saturates the system, i.e. it computes a fixpoint.
+    *
+    * The implementation uses three rewrite rules:
+    *
+    * - `x ~ true` becomes `[x -> true]`.
+    * - `x ~ c` becomes `[x -> c]`.
+    * - `x /\ y /\ ... = true` becomes `[x -> true, y -> true, ...]`.
+    *
+    * For example, if the equation system is:
+    *
+    * {{{
+    *     c1794221043 ~ (x55062 ∧ x55050 ∧ x55046 ∧ x55060 ∧ x55066 ∧ x55040 ∧ x55075 ∧ x55042 ∧ x55058 ∧ x55078)
+    *     x55078 ~ x112431
+    *     c1794221043 ~ x55040
+    *     x55042 ~ x112433
+    *     x55044 ~ x112437
+    *     x55046 ~ (x112435 ∧ x55044)
+    *     x55048 ~ true
+    *     x55050 ~ (x112439 ∧ x55048)
+    *     x55052 ~ x112443
+    *     x55055 ~ true
+    *     x55058 ~ (x55052 ∧ x112441 ∧ x55055)
+    *     x55060 ~ x112446
+    *     x55062 ~ x112448
+    *     x55066 ~ true
+    *     x55075 ~ x112453
+    * }}}
+    *
+    * then after unit propagation it is:
+    *
+    * {{{
+    *     c1794221043 ~ (c1794221043 ∧ x55062 ∧ x55050 ∧ x55046 ∧ x55060 ∧ x55075 ∧ x55042 ∧ x55058 ∧ x55078)
+    *     x55078 ~ x112431
+    *     x55042 ~ x112433
+    *     x55044 ~ x112437
+    *     x55046 ~ (x112435 ∧ x55044)
+    *     x112439 ~ x55050
+    *     x55052 ~ x112443
+    *     x55058 ~ (x55052 ∧ x112441)
+    *     x55060 ~ x112446
+    *     x55062 ~ x112448
+    *     x55075 ~ x112453
+    * }}}
+    *
+    * with the substitution:
+    *
+    * {{{
+    *     x55048 -> true
+    *     x55055 -> true
+    *     x55066 -> true
+    *     x55040 -> c1794221043
+    * }}}
+    *
+    * Note that several equations were simplified.
+    *
+    * Note: We do not propagate false. This extended can be added, if needed.
+    *
+    * Note: We use `subst.extended` to check for conflicts. For example, if we already know that `s = [x -> c17]` and we
+    * learn that `x -> true` then we will try to extend s with the new binding which will raise a [[ConflictException]].
+    */
+  private def propagateUnit(l: List[Equation], s: BoolSubstitution): (List[Equation], BoolSubstitution) = {
+    var pending = l
+    var subst = s
+
+    // We iterate until no changes are detected.
+    var changed = true
+    while (changed) {
+      changed = false
+
+      var rest: List[Equation] = Nil
+      for (e <- pending) {
+        e match {
+          // Case 1: x ~ true
+          case Equation(Term.Var(x), Term.True, loc) =>
```

Comment:
What if
```
x ~ true
x ~ false
```

does that not return `[x -> false]`?

Reply:
Oh I see that `extended` does the checking. I don't know if its wrong, but it feels unexpected to me at least

Reply:
Yes. I added a comment. Will think about whether a refactor would be possible.

Reply:
Added comment here: #7363